test/jdk/valhalla/valuetypes/NullRestrictedArraysTest.java

```diff
@@ -171,65 +172,157 @@ public void testVarHandle() {
-        assertThrows(NullPointerException.class, () -> vh.setVolatile(array, 0, null));
-        assertThrows(NullPointerException.class, () -> vh.setOpaque(array, 0, null));
-        assertThrows(NullPointerException.class, () -> vh.setRelease(array, 0, null));
-
-        assertThrows(NullPointerException.class, () -> vh.compareAndSet(array, 1, value1, null));
-        assertThrows(NullPointerException.class, () -> vh.compareAndExchange(array, 1, value1, null));
-        assertThrows(NullPointerException.class, () -> vh.compareAndExchangeAcquire(array, 1, value1, null));
-        assertThrows(NullPointerException.class, () -> vh.compareAndExchangeRelease(array, 1, value1, null));
-        assertThrows(NullPointerException.class, () -> vh.weakCompareAndSet(array, 1, value1, null));
-        assertThrows(NullPointerException.class, () -> vh.weakCompareAndSetAcquire(array, 1, value1, null));
-        assertThrows(NullPointerException.class, () -> vh.weakCompareAndSetPlain(array, 1, value1, null));
-        assertThrows(NullPointerException.class, () -> vh.weakCompareAndSetRelease(array, 1, value1, null));
-    }
+        // test set with non-null values
+
+        vh.set(array, 0, value1);
+        assertEquals(vh.get(array, 0), value1);
+        vh.setVolatile(array, 0, value1);
+        assertEquals(vh.get(array, 0), value1);
+        vh.setOpaque(array, 0, value1);
+        assertEquals(vh.get(array, 0), value1);
+        vh.setRelease(array, 0, value1);
+        assertEquals(vh.get(array, 0), value1);
+
+        assertTrue(vh.compareAndSet(array, 1, value1, value2));
+        assertEquals(vh.get(array, 1), value2);
+        vh.set(array, 1, value1);
+
+        assertEquals(vh.compareAndExchange(array, 1, value1, value2), value1);
+        assertEquals(vh.get(array, 1), value2);
+        vh.set(array, 1, value1);
+
+        assertEquals(vh.compareAndExchangeAcquire(array, 1, value1, value2), value1);
+        assertEquals(vh.get(array, 1), value2);
+        vh.set(array, 1, value1);
+
+        assertEquals(vh.compareAndExchangeRelease(array, 1, value1, value2), value1);
+        assertEquals(vh.get(array, 1), value2);
+        vh.set(array, 1, value1);
+
+        assertTrue(vh.weakCompareAndSet(array, 1, value1, value2));
+        assertEquals(vh.get(array, 1), value2);
+        vh.set(array, 1, value1);
+
+        assertTrue(vh.weakCompareAndSetAcquire(array, 1, value1, value2));
+        assertEquals(vh.get(array, 1), value2);
+        vh.set(array, 1, value1);
+
+        assertTrue(vh.weakCompareAndSetPlain(array, 1, value1, value2));
+        assertEquals(vh.get(array, 1), value2);
+        vh.set(array, 1, value1);
 
+        assertTrue(vh.weakCompareAndSetRelease(array, 1, value1, value2));
+        assertEquals(vh.get(array, 1), value2);
+        vh.set(array, 1, value1);
+
+        // test atomic set with null witness
+
+        assertFalse(vh.compareAndSet(array, 2, null, value1));
```

Comment:
I realized that the VH code we have is liberal w.r.t. `null` witness values in CAS-like operations -- in the sense that it allows them, even though the underlying variable is null-restricted. For now I decided to preserve this behavior (instead e.g. throwing an IAE)